### PR TITLE
Clean subscriptions' cache for auto-install flow

### PR DESCRIPTION
**Pull request by tgglv**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

It adds clearing cache before getting customer's subscriptions. It will be used for paid (non-WP.org ones) products because they have subscriptions.

Closes https://github.com/Automattic/woocommerce.com/issues/6532.

### How to test the changes in this Pull Request:
This change is used to In-App Purchases, thus to test it holistically we need a dev WCCOM site and a sample Store. To decrease the complexity of testing we could use the minimal scenario.

**What do we need to simulate to check that these changes what we exactly want?**

(In-App Purchases or auto-install flow)
1. A customer connects their store to WCCOM
1. The customer buys a plugin on WCCOM
1. WCCOM asks the Store to run an installation
1. The Store finds out the purchased products
1. If a product that should be installed was purchased, it will be installed

**When does the problem happen?**

When the Store finds out products were purchased it gets customer's subscriptions from the `_woocommerce_helper_subscriptions` transient, and if it's empty, loads them from WCCOM into the transient.
When the transient is already set it could be got stalled just after purchase because it won't be updated without manual clearing. That's why we should clear the transient before asking whether a product was purchased or not.

#### How to test: the easiest way
We'll see that getting subscriptions relies on the transient.
1. Make sure that a store you're using for testing uses this version of WC and it **isn't** connected to test WCCOM (it needs because if you don't have test WCCOM, it would be hard to test on a local Store with the use of real WCCOM).
1. Run in `wp shell` (it should return an empty array and `false`):
```bash
get_transient( '_woocommerce_helper_subscriptions' );
WC_Helper::has_product_subscription( 27147 )
```

3. Let's fill the transient with WC Subscription plugin and ask about a sub again (the last command should return `true`):
```bash
set_transient( '_woocommerce_helper_subscriptions', json_decode('[{"product_key":"W00-050fd691-40c0-94fb-2129ea23348a","product_id":27147,"product_name":"WooCommerce Subscriptions","product_url":"https:\/\/woocommerce.test\/products\/woocommerce-subscriptions\/","product_status":"publish","key_type":"single","key_type_label":"Single site","key_parent_order_item_id":null,"autorenew":false,"connections":[3],"legacy_connections":[],"shares":[],"lifetime":false,"expires":1599609600,"expired":false,"expiring":false,"sites_max":1,"sites_active":1,"maxed":false}]', true) );
WC_Helper::has_product_subscription( 27147 )
```
That's why I added clearing cache before getting customer's subscriptions. 

#### How to test: the E2E way
1. Connect the Store to WCCOM
1. Imagine, you have one plugin (WC Subscriptions), so set your transient's value: 
```bash
$ wp --allow-root shell
wp> set_transient( '_woocommerce_helper_subscriptions', json_decode('[{"product_key":"W00-050fd691-40c0-94fb-2129ea23348a","product_id":27147,"product_name":"WooCommerce Subscriptions","product_url":"https:\/\/woocommerce.test\/products\/woocommerce-subscriptions\/","product_status":"publish","key_type":"single","key_type_label":"Single site","key_parent_order_item_id":null,"autorenew":false,"connections":[3],"legacy_connections":[],"shares":[],"lifetime":false,"expires":1599609600,"expired":false,"expiring":false,"sites_max":1,"sites_active":1,"maxed":false}]', true) );
```

3. Check out the result of setting transient in the WooCommerce.com Subscriptions tab, you should have only `WooCommerce Subscriptions` plugin here:
image

4. Start your purchase from WP Admin > WooCommerce > Extensions > Browse Extensions try to buy any paid plugin other than WooCommerce Subscription, e.g. Product Add-ons (its ID is `18618`)
1. Go through the payment flow and stop before clicking "Add to Site" on test WCCOM's order complete page.
1. At the moment, the Store actually doesn't know that a customer bought an extension. Please comment line with `WC_Helper::_flush_subscriptions_cache();` 
1. Click the "Add to Site" link in a new tab, wait for the result:
missing_product_subscriptions

8. Uncomment `WC_Helper::_flush_subscriptions_cache();` back and click the "Try again now" button
1. Ensure that plugin was installed
1. You could also check the value of transient. In particular, that cache includes Product Add-ons now (also with the use of `wp shell`, it should return `true`):
```bash
0 < count( array_filter( get_transient( '_woocommerce_helper_subscriptions' ), function( $item ){ return 18618 === $item['product_id']; } ));
```

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

Suggested message for a changelog:
```
Fix - Fixed in-app purchase flow's `Missing product subscription` error.
```
More suitable messages are always appreciated.

### CODE REVIEW

File: includes/wccom-site/class-wc-wccom-site-installer.php [hunk 1]
```diff
@@ -293,6 +293,7 @@ private static function get_product_info( $product_id ) {
 		} else {
 			// Retrieve download URL for non-wporg product.
 			WC_Helper_Updater::flush_updates_cache();
+			WC_Helper::_flush_subscriptions_cache();
```
Comment (haszari) [inline includes/wccom-site/class-wc-wccom-site-installer.php]: This is too late - the previous `elseif` is where we check the subscription - see `WC_Helper::has_product_subscription( $product_id )` above.

I'd suggest that we either move this earlier in the function, or explictly clear all the helper caches at the beginning the `POST install` (start) API handling.

**Review by haszari — CHANGES_REQUESTED** [review]
This is a nice clean change :) however ..

I tested this and it didn't fix the issue; the cache needs to be cleared before we do the subscription check. Suggest that we move the cache clear code earlier in `get_product_info()`, or consider clearing these caches explicitly when handling a "start install" request.

File: includes/wccom-site/class-wc-wccom-site-installer.php (hunk 1 above)
Comment (tgglv) [reply includes/wccom-site/class-wc-wccom-site-installer.php]: Indeed, @haszari, I know. Sorry for having `needs review` status. Actually code was in progress. I chose clear the cache after figuring out whether it's WP.org product or not.

**Review by vedanshujain — COMMENTED** [review]
Looks like testing instructions for easy way are a bit outdated after the recent changeset. I tested with mocking `_get_subscriptions_from_product_id` to always return true and product info call to always return 200 and then it only failed at the update step which I expected it to.

So it's looking fine now. I just left a minor question.

**Review by haszari — APPROVED** [review]
Retested, this is looking good - thanks @tgglv :)

@vedanshujain what do we need to do to get this in a release and merged?

File: includes/wccom-site/class-wc-wccom-site-installer.php [hunk 2]
```diff
@@ -287,20 +287,27 @@ private static function get_product_info( $product_id ) {
...
 			$product_info['download_url'] = $result['download_link'];
-		} elseif ( ! WC_Helper::has_product_subscription( $product_id ) ) {
+			return $product_info;
+		}
+
+		// Clear the cache of customer's subscription before asking for them.
+		// Thus, they will be re-fetched from WooCommerce.com.
+		WC_Helper::_flush_subscriptions_cache();
+
+		if ( ! WC_Helper::has_product_subscription( $product_id ) ) {
```
[the remark this answers is not in the document]
Comment (tgglv) [reply includes/wccom-site/class-wc-wccom-site-installer.php]: Indeed, @vedanshujain, that implementation would produce one subscription fetching per every single product, and that isn't good.

So, I moved cache flushing upper here: https://github.com/woocommerce/woocommerce/pull/24810/commits/386aedd73d8b31718a74b3c5514b159a855a5d40

This way, the next attempt to get subscriptions will fetch all updated subscription data once, so we will reach our goal. Also, we don't need to make code more complicated by preparing a list of products for `_update_check` calling.

What do you think about this implementation?

**Comment by haszari** [comment]
@tgglv @vedanshujain @peterfabian

**Pull request merged**